File: 2021 KAKAO BLIND RECRUITMENT/메뉴 리뉴얼.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

map<string, int> list;
map<int, int> cntList;
map<string, int>::iterator listIt;
map<int, int>::iterator cntIt;


void dfs(string order, int idx, int sz, int cnt, string tmp) {
    tmp += order[idx];
    if(cnt == sz) {
        list[tmp]++;
        if(cntList[sz] < list[tmp])
            cntList[sz] = list[tmp];
        return;
    }
    for(int i = idx+1; i < order.size(); ++i)
        dfs(order, i, sz, cnt+1, tmp);
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    for(int i = 0; i < course.size(); ++i) {
        int sz = course[i];
        for(int j = 0; j < orders.size(); ++j) {
            for(int k = 0; k < orders[j].size(); ++k) {
                sort(orders[j].begin(), orders[j].end());
                dfs(orders[j], k, sz, 1, "");
            }
        }
    }
    for(cntIt = cntList.begin(); cntIt != cntList.end(); ++cntIt) {
        cout << cntIt->first << " " << cntIt->second << endl;
        for(listIt = list.begin(); listIt != list.end(); ++listIt) {
            if(cntIt->second >= 2)
                if(listIt->first.size() == cntIt->first && listIt->second == cntIt->second)
                    answer.push_back(listIt->first);
        }
    }
    sort(answer.begin(), answer.end());
    return answer;
}

File: 2021 KAKAO BLIND RECRUITMENT/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    String[] Answer;
    String[] Orders;
    int[] Course; 
    int MaxCnt;
    Map<String, Integer> Comb;
    PriorityQueue<String> PQ;

    public void comb(String order, int idx, int c, String foods) {
        if (foods.length() == c) {
            // System.out.println(foods);
            
            char arr[] = foods.toCharArray();
            Arrays.sort(arr);
            foods = new String(arr);
            
            if (!Comb.containsKey(foods))
                Comb.put(foods, 0);
            
            Comb.put(foods, Comb.get(foods) + 1);
            MaxCnt = Math.max(MaxCnt, Comb.get(foods));
            return;
        }
        
        for (int i = idx; i < order.length(); ++i) {
            comb(order, i + 1, c, foods + order.charAt(i));
        }
    }
    
    public String[] solution(String[] orders, int[] course) {
        Orders = orders;
        Course = course;
        PQ = new PriorityQueue<>();
        
		for (int c : course) {
            Comb = new HashMap<>();
            MaxCnt = 0;
            
            for (String order : orders) {
                comb(order, 0, c, "");
            }
            
 			for (String k : Comb.keySet()) {
                if (Comb.get(k) == MaxCnt && MaxCnt > 1)
                    PQ.offer(k);
            }      
        }
        Answer = new String[PQ.size()];
        int idx = 0;
        while (!PQ.isEmpty()) {
			Answer[idx++] = PQ.poll();
        }
        
        return Answer;
    }
}
